src/debug/telemetry/telemetry2_thread_info.c

```diff
@@ -0,0 +1,349 @@
+/* SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) 2024 Intel Corporation
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <stdio.h>
+#include <soc.h>
+#include <adsp_debug_window.h>
+
+#include "telemetry2_slot.h"
+
+LOG_MODULE_REGISTER(thread_info);
+
+/* to adsp backend header, data to be shared with user space client*/
+#define THREAD_INFO_MAX_THREADS 16
+
+struct thread_info {
+	char name[14];
+	uint8_t stack_usage;	/* precentage */
+	uint8_t cpu_usage;	/* precentage */
+} __packed;
+
+#define THREAD_INFO_VERSION_0_0		0
+
+#define THREAD_INFO_STATE_UNINITIALIZED	0
+#define THREAD_INFO_STATE_BEING_UPDATED	1
+#define THREAD_INFO_STATE_UPTODATE	2
+
+struct thread_info_core {
+	uint8_t state;
+	uint8_t counter;		/* incremented every round */
+	uint8_t load;			/* precentage */
+	uint8_t thread_count;
+	struct thread_info thread[THREAD_INFO_MAX_THREADS];
+} __packed __aligned(CONFIG_DCACHE_LINE_SIZE);
+
+struct thread_info_chunk {
+	struct telemetry2_chunk_hdr hdr;
+	uint16_t core_count;
```

Comment:
a thread can run on multiple cores?

Reply:
This is the header of the whole thread info chunk, and the number indicates how many cores we have. The code keeps all the core separate all the time. Nothing else seems to work (DSP hangs). So there is a thread running on each core reporting the threads on that core. This header is just written once in the beginning and is used by the user-space client to find the core specific sections. I'll add some more comments.